models/GroupsService.php

```diff
@@ -15,130 +15,148 @@
     {
-        $subjectClass = TestTakerService::singleton()->getRootClass();
-        $users = $subjectClass->searchInstances([
-            self::PROPERTY_MEMBERS_URI => $groupUri
-        ], [
-            'recursive' => true, 'like' => false
-        ]);
-        return $users;
+        return $this->getTestTakerService()->getRootClass()->searchInstances(
+            [self::PROPERTY_MEMBERS_URI => $groupUri],
+            ['recursive' => true, 'like' => false]
+        );
     }
-    
+
     /**
-     * Add a User to a Group
-     *
-     * @param string $userUri
-     * @param core_kernel_classes_Resource $group
-     * @return boolean
+     * Adds a user to a Group.
      */
-    public function addUser($userUri, core_kernel_classes_Resource $group)
+    public function addUser(string $userUri, core_kernel_classes_Resource $group): bool
     {
-        $user = new \core_kernel_classes_Resource($userUri);
-        return $user->setPropertyValue(new core_kernel_classes_Property(self::PROPERTY_MEMBERS_URI), $group);
+        return $this->getModel()->getResource($userUri)->setPropertyValue(
+            $this->getModel()->getProperty(self::PROPERTY_MEMBERS_URI),
+            $group
+        );
     }
-    
+
     /**
-     * Remove a User from a Group
+     * Removes a user from a Group.
+     */
+    public function removeUser(string $userUri, core_kernel_classes_Resource $group): bool
+    {
+        return $this->getModel()->getResource($userUri)->removePropertyValue(
+            $this->getModel()->getProperty(self::PROPERTY_MEMBERS_URI),
+            $group
+        );
+    }
+
+    /**
+     * Creates a duplicate of the given group instance into the given class,
+     * copying associations for former Test Takers and Deliveries to the new group.
      *
-     * @param string $userUri
-     * @param core_kernel_classes_Resource $group
-     * @return boolean
+     * @throws common_Exception
+     * @throws common_exception_Error
+     * @throws FileExistsException
      */
-    public function removeUser($userUri, core_kernel_classes_Resource $group)
+    public function cloneInstance(
+        core_kernel_classes_Resource $instance,
+        core_kernel_classes_Class $class = null
+    ): core_kernel_classes_Resource {
+        $newGroup = parent::cloneInstance($instance, $class);
+
+        // Test takers assigned to the group are not copied by the parent class
+        // method (but deliveries assigned to the group are), so we need to
+        // assign them here to the new group.
+        //
+        foreach ($this->getUsers($instance->getUri()) as $user) {
+            $this->addUser($user->getUri(), $newGroup);
+        }
+
+        return $newGroup;
+    }
+
+    private function getTestTakerService(): TestTakerService
+    {
+        return $this->testTakerService ?? TestTakerService::singleton();
+    }
+
+    public function setTestTakerService(TestTakerService $service): void
     {
```

Comment:
so, a setter for a test :thinking: 
beware this is the same as having a param in the constructor for a test, as i said yesterday its okey for me, as it is harmless, but i would use a stub on the test and just add this method there. At least it could be used on getTestTakerService to set it if it was not existing previously